vendor/github.com/cyphar/filepath-securejoin/mkdir_linux.go

```diff
@@ -0,0 +1,229 @@
+		}
+	}()
+	if err != nil && !errors.Is(err, unix.ENOENT) {
+		return nil, fmt.Errorf("find existing subpath of %q: %w", unsafePath, err)
+	}
+
+	// If there is an attacker deleting directories as we walk into them,
+	// detect this proactively. Note this is guaranteed to detect if the
+	// attacker deleted any part of the tree up to currentDir.
+	//
+	// Once we walk into a dead directory, partialLookupInRoot would not be
+	// able to walk further down the tree (directories must be empty before
+	// they are deleted), and if the attacker has removed the entire tree we
+	// can be sure that anything that was originally inside a dead directory
+	// must also be deleted and thus is a dead directory in its own right.
+	//
+	// This is mostly a quality-of-life check, because mkdir will simply fail
+	// later if the attacker deletes the tree after this check.
+	if err := isDeadInode(currentDir); err != nil {
+		return nil, fmt.Errorf("finding existing subpath of %q: %w", unsafePath, err)
+	}
+
+	// Re-open the path to match the O_DIRECTORY reopen loop later (so that we
+	// always return a non-O_PATH handle). We also check that we actually got a
+	// directory.
+	if reopenDir, err := Reopen(currentDir, unix.O_DIRECTORY|unix.O_CLOEXEC); errors.Is(err, unix.ENOTDIR) {
+		return nil, fmt.Errorf("cannot create subdirectories in %q: %w", currentDir.Name(), unix.ENOTDIR)
+	} else if err != nil {
+		return nil, fmt.Errorf("re-opening handle to %q: %w", currentDir.Name(), err)
+	} else {
+		_ = currentDir.Close()
+		currentDir = reopenDir
+	}
+
+	remainingParts := strings.Split(remainingPath, string(filepath.Separator))
+	if slices.Contains(remainingParts, "..") {
+		// The path contained ".." components after the end of the "real"
+		// components. We could try to safely resolve ".." here but that would
+		// add a bunch of extra logic for something that it's not clear even
+		// needs to be supported. So just return an error.
+		//
+		// If we do filepath.Clean(remainingPath) then we end up with the
+		// problem that ".." can erase a trailing dangling symlink and produce
+		// a path that doesn't quite match what the user asked for.
+		return nil, fmt.Errorf("%w: yet-to-be-created path %q contains '..' components", unix.ENOENT, remainingPath)
+	}
+
+	// Make sure the mode doesn't have any type bits.
+	mode &^= unix.S_IFMT
+	// What properties do we expect any newly created directories to have?
+	var (
+		// While umask(2) is a per-thread property, and thus this value could
+		// vary between threads, a functioning Go program would LockOSThread
+		// threads with different umasks and so we don't need to LockOSThread
+		// for this entire mkdirat loop (if we are in the locked thread with a
+		// different umask, we are already locked and there's nothing for us to
+		// do -- and if not then it doesn't matter which thread we run on and
+		// there's nothing for us to do).
+		expectedMode = uint32(unix.S_IFDIR | (mode &^ getUmask()))
```

Comment:
This may cause a regression for `dind` now. @cyphar 
```
@lifubang ➜ ~/ubuntu $ sudo /workspaces/runc/runc create test
ERRO[0000] runc create failed: unable to start container process: error during container init: error mounting "proc" to rootfs at "/proc": possible attack detected: newly created directory "/home/codespace/ubuntu/rootfs/proc" has incorrect mode 0o42755 (expected 0o40755)
```
This is from github codespace online webide, I think it looks like a `dind`, and it works before.

Reply:
Maybe we should consider the SUID bit.

Reply:
I tested MkdirAll with suid bits and it seemed that it didn't work properly so masking it made more sense, let me test it again...

Reply:
But this umask check isn't the problem though -- the umask doesn't affect those bits.

Reply:
Yeah, I noticed this when working on it but assumed that `os.MkdirAll` had the same issue because it seems suid (and sgid) bits aren't set if you pass them to `mkdirat`. Let me take a look.

This check isn't the problem though (unless you want to silently ignore suid bits being dropped) -- we pass the requested mode to `mkdirat` directly so the issue is that `mkdirat` doesn't set them.

Reply:
(Also I don't get why you'd want to set the mode of `/proc` that way. That seems very strange.)

Reply:
No, it’s the default action, I didn’t set anything. It’s also has an error for normal bind mount, if the destination dir is not existing in the container rootfs.

Reply:
See from the result, I think ‘mkdirat’ has honored the suid bit even though there is no suid bit passed in. But I don’t know where is the default suid bit come from, maybe from the parent dir?

Reply:
https://github.com/cyphar/filepath-securejoin/pull/25 should fix the issue, I'll send a PR with it applied.